2020-hold-my-context.md

```diff
@@ -0,0 +1,58 @@
+# Hold my context!
+
+☝️  _the presentation title_☝️
+
+## Short summary for display
+
+Preserving contextual data across asynchronous calls have been always tricky
+in Node.js land. For example, how to pass information about the current user
+or a correlation (transaction) id from the code processing incoming HTTP
+requests to the code calling a backend web service? Traditional runtimes like
+C and Java can utilize thread-local storage, which does not work for Node's
+single-threaded event loop.
+
+The talk will describe different solutions people have been using (and
+abandoning) over the past 6+ years, dive into history of domains and async
+hooks, explain what's so difficult about context keeping, and finally offer
+bright future that's coming with the recently introduced `AsyncLocalStorage` API.
+
+## Talk description and abstract
+
+_Structure, table of contents and main content points._
+
+1. Why would anyone need to preserve a context? Start with an example explaining
+  the problem faced by server developers: how to forward correlaction-id from
+  an incoming request to a backend web service. Show how easy this is in sync
+  world (C/Java/.NET).
+
+2. Solution 1: core `domains`. While they can track async transitions and
+  continuations, they are not meant to store custom data.
+
+3. Solution 2: `continuation-local-storage`. The go-to solution during Node.js
+   0.10 days. Unfortunately, it often drops the context and requires
+   applications to cooperate, which often means fixing dependencies 😱.
+
+4. Explain why is it difficult to preserve the context, a.k.a. "connection pool
+   problem". Talk about our own experience with issues caused by MongoDB driver.
+   Mention other use cases affected by this problem (APM tools like NewRelic).
```

Comment:
Promises are a huge problem as well. Google's native promises have support for async hooks, forced to be added by Node.js, but thirdparty promises don't, so any non-native thenable breaks the async continuation. Apparently,  its particularly common for DB drivers.